crates/exex/exex/src/backfill/stream.rs

```diff
@@ -0,0 +1,99 @@
+use super::job::SingleBlockBackfillJob;
+use futures::{
+    stream::{FuturesOrdered, Stream},
+    StreamExt,
+};
+use reth_evm::execute::{BlockExecutionError, BlockExecutionOutput, BlockExecutorProvider};
+use reth_primitives::{BlockNumber, BlockWithSenders, Receipt};
+use reth_provider::{BlockReader, HeaderProvider, StateProviderFactory};
+use std::{
+    ops::RangeInclusive,
+    pin::Pin,
+    task::{ready, Context, Poll},
+};
+use tokio::task::JoinHandle;
+
+type BackfillTasks = FuturesOrdered<
+    JoinHandle<Result<(BlockWithSenders, BlockExecutionOutput<Receipt>), BlockExecutionError>>,
+>;
+
+/// The default capacity of active tasks for [`BackFillJobStream`].
+const DEFAULT_CAPACITY: usize = 100;
+
+/// Stream for processing backfill jobs asynchronously.
+///
+/// This struct manages the execution of [`SingleBlockBackfillJob`] tasks, allowing blocks to be
+/// processed asynchronously within a specified range.
+#[derive(Debug)]
+pub struct BackFillJobStream<E, P> {
+    job: SingleBlockBackfillJob<E, P>,
+    tasks: BackfillTasks,
+    range: RangeInclusive<BlockNumber>,
+    capacity: usize,
+}
+
+impl<E, P> BackFillJobStream<E, P>
+where
+    E: BlockExecutorProvider + Clone + Send + 'static,
+    P: HeaderProvider + BlockReader + StateProviderFactory + Clone + Send + 'static,
+{
+    /// Creates a new [`BackFillJobStream`] with the default capacity size.
+    ///
+    /// # Parameters
+    /// - `job`: The [`SingleBlockBackfillJob`] to be executed asynchronously.
+    ///
+    /// # Returns
+    /// A new instance of [`BackFillJobStream`] with the default capacity size.
+    pub fn new(job: SingleBlockBackfillJob<E, P>) -> Self {
+        Self::new_with_capacity(job, DEFAULT_CAPACITY)
+    }
+
+    /// Creates a new [`BackFillJobStream`] with a specified capacity size.
+    ///
+    /// # Parameters
+    /// - `job`: The [`SingleBlockBackfillJob`] to be executed asynchronously.
+    /// - `capacity`: The capacity of the [`BackFillJobStream`] to handle active tasks.
+    ///
+    /// # Returns
+    /// A new instance of [`BackFillJobStream`] with the specified capacity size.
+    pub fn new_with_capacity(job: SingleBlockBackfillJob<E, P>, capacity: usize) -> Self {
```

Comment:
we also need a setting and a `fn with(self)-> Self` helper so you can call into_stream().with_parallelism()